config/reminder-config.yaml.example

```diff
@@ -0,0 +1,42 @@
+#
+# Copyright 2024 Stacklok, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+recurrence:
+  interval: "1h"
+  batch_size: 100
+  max_per_project: 10
+  min_project_fetch_limit: 10
+  min_elapsed: "1h"
+
+database:
+  dbhost: "postgres"
+  dbport: 5432
+  dbuser: postgres
+  dbpass: postgres
+  dbname: minder
+  sslmode: disable
+
+cursor_file: "/tmp/reminder-cursor"
+
+logging_level: "debug"
+
+events:
+  sql_connection:
+    dbhost: "reminder-event-postgres"
+    dbport: 5432
+    dbuser: postgres
+    dbpass: postgres
+    dbname: reminder
+    sslmode: disable
```

Comment:
Nope, it's not like that. Minder message queue system would be untouched. We will just add a new subscriber to the server for reminder events. Minder can continue to run using `go-channel`. Minder would connect to reminder using separate config.

Reply:
Yes, it is a standalone publishing system for reminder that minder would subscribe to.

Reply:
> Does this mean that we'll have two different messaging SQL databases

Precisely. But now that you mention it like that, would it be better to have another table (if possible) in the existing database? I'm torn between having another db (maybe 1 is enough) v/s ease of maintenance decoupling the services  🤔

Reply:
I think fewer databases is likely to be easier to manage.  Given the way that watermill-SQL operates, I'm not sure whether it can create different additional tables in the same database, but it should be possible to use different topics if you wish.

I some some skepticism that generating an event in Reminder on one message bus that Minder then needs to watch and convert to a second message on the existing Minder message bus will really have the decoupling effect that you're looking for -- it seems like we'll be adding the complexity of connecting to more databases to Minder, and overall adding configuration and moving parts vs having Reminder use the _same_ Watermill message bus.

Reply:
Now that I think about it (and having written this), it occurs to me that we could _also_ use Watermill to store our cursor state, though we may need to be somewhat careful about not busy-looping in between remediation cycles.

i.e. when reminder starts up, it tries to read a "continue remediations from" message from a Watermill topic.  If there are no messages, it starts at the beginning.  At the end of each iteration, it enqueues a new "continue remediation from (end of batch)" message, followed by retiring the continue message with the start of the batch.

This would also work fine with other messaging systems, as long as they allow for a moderately-long message processing interval and allow queueing a message for later delivery.

Reply:
Yes, an incoming message from reminder would be converted to a reconciliation message and sent to the different message bus i.e. regular minder event processing queue. Using the approach you suggest i.e. having a single message queue so it is easier to maintain it and message doesn't have to hop from message queue a to b.

But still, reminder would have a different topic, so now reminder's message would have to be converted to a message for minder (both of them being in the same messaging queue).

Also, this wouldn't allow reminder to run with gochannel message queue (in minder server).

Reply:
Coming onto the point of storing cursor state in message queue. I don't see any significant advantage, rather making a db query to get the next cursors would increase the processing time. This would also increase the complexity of the logic IMO. Does storing cursor state in db provide any significant advantage over storing in memory (and a file)?

Reply:
Two benefits of storing the state remotely in the database over storing in a file are:

1. We could theoretically run multiple `reminder` servers which would interoperate with each other.
2. We avoid needing to fuss with persistent filesystem storage when deploying on Kubernetes / docker-compose.  docker-compose is probably easier, but on Kubernetes using filesystem storage either means bringing something like NFS into the picture, or using a StatefulSet instead of a Deployment.  StatefulSet updates are a bit trickier to manage high-availability rollouts for compared with Deployments, as you can't have the StatefulSet "surge" (run extra instances) when rolling out -- you need to stop the old instance then try starting the new one.

Reply:
With respect to not being able to use gochannel on Minder with reminder -- I'm not sure I see the benefit of being able to use gochannel for _some_ events if you still need to set up watermill-sql for other events.  Once you're using watermill-sql, you might as well use it for everything...

And my suggestion was to use the same topic as Minder for the published events, so we didn't end up with "reminder events" that needed to be copied to "minder events".

Reply:
Hmm, interesting. 

> We could theoretically run multiple reminder servers which would interoperate with each other.

I can see this happening, but I don't have the design pattern for it, and it isn't required now, so I won't brainstorm on that side. 

Now that we want to store cursor state in the main db, I was still thinking of:

1. Reading complete cursor state from db to memory (during start)
2. Update cursor state in memory then update cursor state in db
3. If update to db fails, rollback changes in memory to be in sync

This approach is likely to change if we want to run multiple reminders simultaneously. This is a cache, and running multiple reminders, we'll need to invalidate/update it frequently, but again, that is a problem of the future. 

(You good with storing in memory for now (cache) ?)

Reply:
Adding cursors to the db would look something like:

```sql
CREATE TABLE reminder_project_cursor (
  id SERIAL PRIMARY KEY, -- Some sort of ID to differentiate between different reminders
  cursor TEXT NOT NULL
);

CREATE TABLE reminder_repository_cursor (
  id SERIAL PRIMARY KEY, -- Some sort of ID to differentiate between different reminders
  project_id UUID NOT NULL,
  provider TEXT NOT NULL,
  cursor TEXT NOT NULL,
  FOREIGN KEY (project_id, provider) REFERENCES providers(project_id, name) ON DELETE CASCADE,
  UNIQUE (project_id, provider)
);
```

Using this approach every reminder service would have some unique identifier associated with it. I had a few concerns:

1. Are we simplifying by storing in db or complicating it more than required? (With in-memory cache and all that stuff)
2. Do we even need to run multiple reminder instances? You told me a lot of other things would break before we scale. 
4. If yes, wouldn't it be simpler to have a reminder service operate on shard? Or set a limit somehow? (Taking a wild guess here)
3. Are we over-engineering this 😬 ?

Reply:
> Hmm, interesting.
> 
> > We could theoretically run multiple reminder servers which would interoperate with each other.
> 
> I can see this happening, but I don't have the design pattern for it, and it isn't required now, so I won't brainstorm on that side.

The idea would be that we could run multiple servers for high availability.  I agree that it's not necessary for right now, as long as we make sure that it's not strictly required to only run one Reminder for correctness.

> (You good with storing in memory for now (cache) ?)

Yes.

Reply:
> Also, do we want to modify `type RepoReconcilerEvent struct` to have a source string field for visibility and helping to differentiate whether the event was generated by reminder or minder. (For topic: `TopicQueueReconcileRepoInit`)

Rather than putting it in the `type RepoReconcilerEvent struct`, can you put something like `periodic=true` or `trigger=revisit` in the message _metadata_?

Ideally, we'd have a good split between routing metadata (message subject, event type, etc) and the message _contents_ (diff of the change, for example).  Having the metadata in the outer envelope allows us to generically do things like 2-lane queues (handle revisits at lower priority than live notifications) and record success rates for revisits separately.  The generic top-level handling layer _should not_ know about the payload schema -- that's for the actual endpoint to decode.

Reply:
> Adding cursors to the db would look something like:
> 
> ```sql
> CREATE TABLE reminder_project_cursor (
>   id SERIAL PRIMARY KEY, -- Some sort of ID to differentiate between different reminders
>   cursor TEXT NOT NULL
> );
> 
> CREATE TABLE reminder_repository_cursor (
>   id SERIAL PRIMARY KEY, -- Some sort of ID to differentiate between different reminders
>   project_id UUID NOT NULL,
>   provider TEXT NOT NULL,
>   cursor TEXT NOT NULL,
>   FOREIGN KEY (project_id, provider) REFERENCES providers(project_id, name) ON DELETE CASCADE,
>   UNIQUE (project_id, provider)
> );
> ```
> 
> Using this approach every reminder service would have some unique identifier associated with it. I had a few concerns:
> 
> 1. Are we simplifying by storing in db or complicating it more than required? (With in-memory cache and all that stuff)
> 2. Do we even need to run multiple reminder instances? You told me a lot of other things would break before we scale.
> 3. If yes, wouldn't it be simpler to have a reminder service operate on shard? Or set a limit somehow? (Taking a wild guess here)
> 4. Are we over-engineering this 😬 ?

Do we need fairness at both the project and the repo level?  My naive thought would be to iterate across all the entities -- projects which had more repos registered would get more "revisits / day" than smaller projects, but that seems appropriate.  We probably want to limit how many concurrent reconciliations are happening in a project, but that might make more sense to do at the Minder level, since we could have a project that genuinely gets changes to 50 repos at the same time.

It's hard to guarantee fairness until we know what fairness looks like.  😁 


In terms of running multiple reminder instances, my goal is that we should be able to upgrade Reminder as follows:

1. Reminder v.123 is running one instance
2. We start a Reminder v.124
3. We shut down Reminder v.123

This is easy to do with a Kubernetes Deployment with `maxSurge: 1`, and avoids getting into a situation where v.124 doesn't work, but we've corrupted state needed for v.123 to operate, so we end up needing to do manual maintenance.  This is also the model we use with the core Minder binary, so it's less surprising than having a different model.  As long as we can use that manner of deployment, I'm happy with many different solutions:

a. Iterate over the database several times faster than the reminder interval, but honor some pacing on actually sending the events.  In this case, we might not need to keep a cursor at all.
b. Add a database table to track the Reminder cursors.
c. Use "continuation messages" in Watermill to describe what we just processed.  e.g. at the end of processing up to row N, send a "now process >N" message, then acknowledge the "now process >N-10" message that was in queue earlier.  If there's no message, start at a random point.

Reply:
> Do we need fairness at both the project and the repo level? My naive thought would be to iterate across all the entities -- projects which had more repos registered would get more "revisits / day" than smaller projects, but that seems appropriate. We probably want to limit how many concurrent reconciliations are happening in a project, but that might make more sense to do at the Minder level, since we could have a project that genuinely gets changes to 50 repos at the same time.
>
> It's hard to guarantee fairness until we know what fairness looks like. 😁

Exactly. My main idea behind this was to ensure fair sharing among projects. If we reconcile large number of entities in a single project, we might hate the rate limits (unprobable, but possible). Current model (which is yet to be pushed) will work on the logic of picking min number of repos form a project (to ensure some fair share), and picking additional repos if there are more spots. 

---

Let's discuss cursors in a discord thread. I also would like reminder to run as a deployment rather than a stateful set.

Reply:
Discord thread: https://discord.com/channels/1184987096302239844/1185287949240242258/1232246213005283348